### PR TITLE
LSP server

### PR DESCRIPTION
**Pull request by 0xGeorgii**
<h3>Confidence Score: 5/5</h3>

The PR is safe to merge — it adds a substantial new LSP server and IDE stack with no regressions to existing code paths.

All core paths — URI conversion, position encoding, diagnostics publishing, the shutdown handshake, and the VS Code lifecycle queue — are correctly implemented and well-covered by tests. The one behavioral edge case (inlay hints falling back to the whole file when a viewport range extends past the last line) is harmless in practice and does not affect correctness for any standard editor flow.

apps/lsp/src/handlers.rs — the inlay_hint range fallback is the only area worth a follow-up look before the feature ships to end users.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| apps/lsp/src/server.rs | New: single-threaded message loop with inline shutdown handshake, request dispatch, and dependent-document republish logic; well-tested with unit coverage of all notification paths. |
| apps/lsp/src/handlers.rs | New: per-request and per-notification handlers; inlay_hint silently falls back to returning all hints when the viewport range is out of bounds (text_range returns None). |
| apps/lsp/src/uri.rs | New: hand-rolled file URI ↔ path conversion with explicit Windows drive-letter handling; comprehensive round-trip tests including percent-encoding, backslash, and query/fragment rejection. |
| apps/lsp/src/convert.rs | New: byte-offset ↔ LSP position/range conversion and IDE→LSP type mappings; every variant is covered explicitly and tested. |
| ide/ide-db/src/database.rs | New: RootDatabase with closure-aware invalidation; open/change/close correctly handles the newly-available widening case for missing imports. |
| ide/ide-db/src/hit_test.rs | New: file-scoped AST hit-testing that descends to the smallest covering node; correctly guards against zero-width default locations claiming offset 0. |
| ide/ide-db/src/analysis.rs | New: FileAnalysis computes type-check, analysis rules, and per-closure file metadata once; closure_paths always includes the entry path to prevent permanent cache poisoning. |
| ide/vfs/src/lib.rs | New: path-interner plus in-memory overlay; no I/O, idempotent interning, correct overlay lifecycle. |
| editors/vscode/src/lsp/client.ts | New: serialized LSP lifecycle management via a promise queue; doStop sets client=undefined before awaiting stop so re-entry is safe; correctly disposes the client on deactivation. |
| editors/vscode/src/lsp/resolve.ts | New: pure binary-resolution helper (settings → managed path → PATH); vscode-free so it is directly unit-testable; correctly refuses to fall back when an explicit path is non-executable. |
| ide/ide/src/lib.rs | New: AnalysisHost/Analysis facade exposing all IDE features by path; closure_line_index reuses the document's already-computed analysis for cross-file position conversion. |
| apps/lsp/src/main.rs | Modified: spawns the server loop on a 64 MiB stack thread (mirrors rust-analyzer) to prevent deep-AST stack overflows on pathological inputs. |

<h3>Sequence Diagram</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Editor as VS Code Editor
    participant Client as LSP Client (client.ts)
    participant Server as inference-lsp (server.rs)
    participant Handler as handlers.rs
    participant IDE as AnalysisHost (ide/src/lib.rs)
    participant DB as RootDatabase (ide-db/src/database.rs)
    participant VFS as Vfs (vfs/src/lib.rs)

    Editor->>Client: activate extension
    Client->>Server: initialize (stdio)
    Server-->>Client: capabilities (FULL sync, hover, def, completion, symbols, hints)

    Editor->>Client: didOpen(uri, text)
    Client->>Server: textDocument/didOpen
    Server->>Handler: did_open(params)
    Handler->>IDE: open_document(path, text)
    IDE->>DB: open_document(path, text)
    DB->>VFS: set_contents(id, text)
    DB->>DB: invalidate dependent analyses
    Handler->>IDE: analysis().diagnostics(path)
    IDE->>DB: analysis(path) compute and memoize FileAnalysis
    DB-->>IDE: FileAnalysis
    IDE-->>Handler: Vec Diagnostic
    Handler-->>Server: PublishDiagnosticsParams plus dependents
    Server->>Client: textDocument/publishDiagnostics x N open docs

    Editor->>Client: hover request
    Client->>Server: textDocument/hover
    Server->>Handler: hover(params)
    Handler->>IDE: analysis().line_index(path)
    Handler->>IDE: analysis().hover(path, offset)
    IDE-->>Handler: Option Hover
    Handler-->>Server: Option Hover
    Server-->>Client: result

    Editor->>Client: didClose(uri)
    Client->>Server: textDocument/didClose
    Server->>Handler: did_close(params)
    Handler->>IDE: close_document(path)
    IDE->>VFS: remove_contents(id)
    Handler-->>Server: empty PublishDiagnosticsParams
    Server->>Client: textDocument/publishDiagnostics clear diagnostics
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Editor as VS Code Editor
    participant Client as LSP Client (client.ts)
    participant Server as inference-lsp (server.rs)
    participant Handler as handlers.rs
    participant IDE as AnalysisHost (ide/src/lib.rs)
    participant DB as RootDatabase (ide-db/src/database.rs)
    participant VFS as Vfs (vfs/src/lib.rs)

    Editor->>Client: activate extension
    Client->>Server: initialize (stdio)
    Server-->>Client: capabilities (FULL sync, hover, def, completion, symbols, hints)

    Editor->>Client: didOpen(uri, text)
    Client->>Server: textDocument/didOpen
    Server->>Handler: did_open(params)
    Handler->>IDE: open_document(path, text)
    IDE->>DB: open_document(path, text)
    DB->>VFS: set_contents(id, text)
    DB->>DB: invalidate dependent analyses
    Handler->>IDE: analysis().diagnostics(path)
    IDE->>DB: analysis(path) compute and memoize FileAnalysis
    DB-->>IDE: FileAnalysis
    IDE-->>Handler: Vec Diagnostic
    Handler-->>Server: PublishDiagnosticsParams plus dependents
    Server->>Client: textDocument/publishDiagnostics x N open docs

    Editor->>Client: hover request
    Client->>Server: textDocument/hover
    Server->>Handler: hover(params)
    Handler->>IDE: analysis().line_index(path)
    Handler->>IDE: analysis().hover(path, offset)
    IDE-->>Handler: Option Hover
    Handler-->>Server: Option Hover
    Server-->>Client: result

    Editor->>Client: didClose(uri)
    Client->>Server: textDocument/didClose
    Server->>Handler: did_close(params)
    Handler->>IDE: close_document(path)
    IDE->>VFS: remove_contents(id)
    Handler-->>Server: empty PublishDiagnosticsParams
    Server->>Client: textDocument/publishDiagnostics clear diagnostics
```

</a>

<sub>Reviews (2): Last reviewed commit: ["more tests"](https://github.com/inferara/inference/commit/ff43bfe0995ff72f3fe9992bbfd07f142f207e6d) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44710700)</sub>

**Context used:**

- Rule used - When reviewing LSP server code in this repo, prefe... ([source](https://app.greptile.com/inferara/github/Inferara/inference/-/custom-context?memory=a7d48768-cc12-4216-9285-c41cf54b1d47))

### CODE REVIEW

File: apps/lsp/src/convert.rs [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+        ide::SymbolKind::Function => SymbolKind::FUNCTION,
+        ide::SymbolKind::Struct => SymbolKind::STRUCT,
+        ide::SymbolKind::Enum => SymbolKind::ENUM,
+        ide::SymbolKind::EnumVariant => SymbolKind::ENUM_MEMBER,
+        ide::SymbolKind::Field => SymbolKind::FIELD,
+        ide::SymbolKind::Method => SymbolKind::METHOD,
+        // A spec is a set of laws over a type, closest to an interface/trait.
+        ide::SymbolKind::Spec => SymbolKind::INTERFACE,
+        ide::SymbolKind::Constant => SymbolKind::CONSTANT,
+        ide::SymbolKind::TypeAlias => SymbolKind::TYPE_PARAMETER,
```
[the remark this answers is not in the document]
Comment (0xGeorgii) [reply apps/lsp/src/convert.rs:84]: Rejecting this suggestion: rust-analyzer — the precedent our `ide/` stack is modeled on — maps `TypeAlias` to `TYPE_PARAMETER` (see `symbol_kind` in `crates/rust-analyzer/src/lsp/to_proto.rs`: `SymbolKind::TypeAlias | SymbolKind::TypeParam | SymbolKind::SelfType => lsp_types::SymbolKind::TypeParameter`). LSP has no dedicated kind for type aliases, so every server picks a stand-in; our primary audience already sees the `TYPE_PARAMETER` icon for type aliases in rust-analyzer daily, making it the more familiar choice. Please remember for next time: when a mapping decision is a judgment call, we follow rust-analyzer conventions.

File: apps/lsp/src/handlers.rs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+pub(crate) fn did_change(
+    state: &mut ServerState,
+    params: DidChangeTextDocumentParams,
+) -> Option<PublishDiagnosticsParams> {
+    let uri = params.text_document.uri;
+    let version = params.text_document.version;
+    let path = uri::to_path(&uri)?;
+    // Full-text sync: the last content change carries the whole new document.
+    let text = params.content_changes.into_iter().next_back()?.text;
```
[the remark this answers is not in the document]
Comment (0xGeorgii) [reply apps/lsp/src/handlers.rs:171]: Rejecting this suggestion: the LSP spec defines `contentChanges` as a sequence applied in order. With FULL sync a conforming client sends exactly one change — where `next()` and `next_back()` are equivalent — but if a client ever batches multiple full-text changes into one notification, the last one is the final document state; taking the first would silently apply stale text. `next_back()` is the defensively correct choice, so we're keeping it. Please remember for next time: we prefer defensive handling of nonconforming LSP clients when it costs nothing.
